bitcoin/listeners/transaction_listener.go

```diff
@@ -0,0 +1,169 @@
+					if r.Txid == hex.EncodeToString(cb.Txid) {
+						return
+					}
+				}
+				orderId, err := calcOrderId(contract.BuyerOrder)
+				if err != nil {
+					return
+				}
+				log.Notice(funded)
+				if !funded {
+					requestedAmount := int64(contract.VendorOrderConfirmation.RequestedAmount)
+					if funding >= requestedAmount {
+						funded = true
+						l.db.Sales().Put(orderId, *contract, pb.OrderState_FUNDED, false)
+						l.adjustInventory(contract)
+
+						n := notifications.Serialize(
+							notifications.OrderNotification{
+								contract.VendorListings[0].Item.Title,
+								contract.BuyerOrder.BuyerID.Guid,
+								contract.BuyerOrder.BuyerID.BlockchainID,
+								contract.VendorListings[0].Item.Images[0].Hash,
+								int(contract.BuyerOrder.Timestamp.Seconds),
+							})
+
+						l.broadcast <- n
+					}
+				}
+				record := spvwallet.TransactionRecord{
+					Txid:  hex.EncodeToString(cb.Txid),
+					Index: output.Index,
+					Value: output.Value,
+				}
+				records = append(records, record)
+				l.db.Sales().UpdateFunding(orderId, funded, records)
+			}
+		} else {
+			_, addrs, _, _ := txscript.ExtractPkScriptAddrs(output.ScriptPubKey, l.params)
+			contract, _, funded, records, err := l.db.Purchases().GetByPaymentAddress(addrs[0])
+			if err == nil {
+				funding := output.Value
+				for _, r := range records {
+					funding += r.Value
+					// If we've already seen this transaction for some reason, just return
+					if r.Txid == hex.EncodeToString(cb.Txid) {
+						return
+					}
+				}
+				orderId, err := calcOrderId(contract.BuyerOrder)
+				if err != nil {
+					return
+				}
+				if !funded {
+					requestedAmount := int64(contract.BuyerOrder.Payment.Amount)
+					if funding >= requestedAmount {
+						funded = true
+						l.db.Purchases().Put(orderId, *contract, pb.OrderState_FUNDED, true)
+					}
+				}
+				record := spvwallet.TransactionRecord{
+					Txid:  hex.EncodeToString(cb.Txid),
+					Index: output.Index,
+					Value: output.Value,
+				}
+				records = append(records, record)
+				l.db.Purchases().UpdateFunding(orderId, funded, records)
+			}
+		}
+	}
+}
+
+func (l *TransactionListener) adjustInventory(contract *pb.RicardianContract) {
+	inventory, err := l.db.Inventory().GetAll()
+	if err != nil {
+		return
+	}
+	for _, item := range contract.BuyerOrder.Items {
+		var variants []string
+		for _, option := range item.Options {
+			variants = append(variants, option.Value)
+		}
+		for path, c := range inventory {
+			contains := true
+		vi:
+			for i := 0; i < len(variants); i++ {
+				if !strings.Contains(path, variants[i]) {
+					contains = false
+					break vi
+				}
+			}
+			if contains && c > 0 {
+				q := int(item.Quantity)
+				if c-q < 0 {
+					q = 0
+					orderId, err := calcOrderId(contract.BuyerOrder)
+					if err != nil {
+						continue
+					}
+					log.Warning("Order %s purchased more inventory for %s than we have on hand", orderId, path)
```

Comment:
We check the inventory level when the order is placed, but don't increment it until the payment comes through. It's possible we could over sell our inventory this way. The only alternative would be to increment when the order comes in but that opens your node up to an attack where the other side sends bogus orders to make your node tell everyone you're out of stock.